libsolidity/ast/Types.h

```diff
@@ -694,11 +692,30 @@ class BoolType: public Type
 	TypeResult interfaceType(bool) const override { return this; }
 };
 
+/**
+ * Base class for types which consist of other types.
+ */
+class CompositeType: public Type
+{
+protected:
+	CompositeType() = default;
+
+public:
+	/// @returns list of all types the composite type consists of, directly or indirectly,
+	/// including the type itself, without duplicates
+	std::vector<Type const*> fullDecomposition() const;
+
+protected:
+	/// @returns list of types the composite type directly consists of
+	/// (base type for array, members for structs, value type for mappings), might contain duplicates
+	virtual std::vector<Type const*> decomposition() const = 0;
```

Comment:
```suggestion
	/// @returns List of types of the elements that taken together form the composite type
	/// (base type for arrays, member types for structs, value type for mappings, etc.).
	/// Might include duplicates.
	virtual std::vector<Type const*> decomposition() const = 0;
```

Reply:
Well, as @chriseth just pointed out, this is *not* the case :-D.

Reply:
I know, it's simplified and ignores stuff like length field for dynamic arrays. How would you suggest to reword it?

Reply:
Hmmmm...
```
@returns a list of types that together make up the data part of this type. Contains all types that will have to be implicitly stored, whenever an object of this type is stored. In particular, it returns the base type for arrays and array slices, the member types for structs, the component types for tuples and the value type for mappings (note that the key type of a mapping is *not* part of the list).
```
or something in that general area maybe :-)?

Reply:
Done.